### PR TITLE
Game state

### LINKED ISSUE #4: Lobbies have a game state that has a deck and a board.

### PR DESCRIPTION
**Pull request by connorlay**
resolves #4 

Lobbies have a Game, which has a deck and board of cards. Users can create a new game after joining a lobby.

pr3 1
pr3 2

### CODE REVIEW

File: app/controllers/api/v1/application_controller.rb [hunk 1]
```diff
@@ -25,9 +25,13 @@ def authentication_error
...
     end
 
-    def record_not_found_error
+    def existence_error
       render json: { error: I18n.t('errors.404') }, status: 404
     end
 
+    def conflict_error
+      render json: { error: I18n.t('errors.409') }, status: 409
+    end
```
Comment (efatsi) [inline app/controllers/api/v1/application_controller.rb]: some would argue you should only have methods in `ApplicationController` that are shared among more than one other controller. Looks like this isn't used outside of `GamesController` at the moment, maybe others in the future? I leave the decision to you if you want to move it.

File: app/controllers/api/v1/application_controller.rb [hunk 2]
```diff
@@ -25,9 +25,13 @@ def authentication_error
       render json: { error: I18n.t('errors.401') }, status: 401
     end
 
-    def record_not_found_error
+    def existence_error
```
Comment (ltk) [inline app/controllers/api/v1/application_controller.rb]: seems less descriptive/conventional than `record_not_found`. why the change?

File: app/controllers/api/v1/games_controller.rb [hunk 3]
```diff
@@ -1,16 +1,26 @@
 module Api::V1
   class GamesController < ApplicationController
 
+    before_action :conflict_error, if: -> { lobby.game.present? },
```
Comment (ltk) [inline app/controllers/api/v1/games_controller.rb]: semantically, I'm not sure 409 is a perfect fit here. Seems like 405 or just 400 might be more accurate.

File: app/controllers/api/v1/games_controller.rb [hunk 4]
```diff
@@ -1,16 +1,26 @@
...
 
     def show
-      game = Game.find(params[:id])
       render json: game, serializer: GameSerializer
     end
 
+    private
+
+    def game
+      @game ||= Game.find(params[:id])
```
Comment (ltk) [inline app/controllers/api/v1/games_controller.rb]: any authorization for viewing games? Should I be able to see game data for games that I have no part in?

File: app/controllers/api/v1/lobbies_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+module Api::V1
+  class LobbiesController < ApplicationController
+
+    def create
+      lobby = Lobby.create
```
Comment (ltk) [inline app/controllers/api/v1/lobbies_controller.rb]: I often find it confusing to name a local the same name as a method. I often prefer to use `new_lobby` as the variable name in `create` actions if I have a `lobby` method defined elsewhere.

File: app/models/game.rb [hunk 6]
```diff
@@ -1,10 +1,35 @@
...
+  validates :lobby, presence: true, uniqueness: true
 
-  def add_user(user)
-    memberships.create user: user
+  belongs_to :lobby
+
+  before_create :setup_game
+
+  CARD_ATTRIBUTES = {
+    symbol:  [ :diamond, :sguiggle, :oval ],
```
Comment (ltk) [inline app/models/game.rb]: should be `squiggle`?

File: lib/dealer.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  def initialize(attributes)
+    @attributes = attributes
+  end
+
+  def new_deck
+    cards.shuffle
+  end
+
+  def get_card(id)
+    cards[id]
```
Comment (efatsi) [inline lib/dealer.rb]: since you're storing the `id` on all your Card objects, it might be worth doing a [detect](http://apidock.com/ruby/Enumerable/detect) on the id itself, instead of just cueing off the position of a card in the array. If not, there's no reason to store the "id" on Card.

File: app/models/game.rb [hunk 8]
```diff
@@ -1,10 +1,35 @@
...
+    number:  [ :one, :two, :three ],
+    color:   [ :red, :green, :purple ],
+    shading: [ :solid, :striped, :open ]
+  }
+
+  def deck
```
Comment (ltk) [inline app/models/game.rb]: any way to avoid having conflicts between the activerecord method `deck` and this method? this can get pretty confusing pretty quickly for others.

File: lib/card.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+class Card
+
+  attr_reader :id
+
+  def initialize(id, attributes)
+    @id = id
+    @attributes = attributes
+  end
+
+  def method_missing(method_name, *args)
```
Comment (ltk) [inline lib/card.rb]: this is kind of opaque. uses of `method_missing` overrides can be really helpful, or just confusing and impossible to debug. Why'd you choose to use it here?

File: lib/dealer.rb (hunk 7 above)
Comment (efatsi) [inline lib/dealer.rb]: huh, am I misreading this perhaps? I see in your `Game` class you're calling `get_cards` with specific card ids (since `deck` and `board` are really just arrays of card ids right?). But `cards[id]` looks like a regular array index lookup.

File: app/models/game.rb [hunk 10]
```diff
@@ -1,10 +1,35 @@
...
+
+  private
+
+  def dealer
+    @dealer ||= Dealer.new(CARD_ATTRIBUTES)
+  end
+
+  def setup_game
+    self.deck  = dealer.new_deck.map(&:id)
+    self.board = read_attribute(:deck).pop(12)
```
Comment (efatsi) [inline app/models/game.rb]: you can store arrays of numbers in an integer field?

File: spec/requests/games_spec.rb [hunk 11]
```diff
@@ -2,24 +2,34 @@
...
-        expect(json['data']['attributes']).to have_key("created_at")
+      end
+      it "responds with card data" do
+        expect(json['data']['attributes']).to have_key("board")
+        expect(json['data']['attributes']).to have_key("deck")
+        expect(json['data']['attributes']['board'].all? {
+          |card| card['attributes'].keys == [ "symbol", "number", "color", "shading" ]
+        }).to be true
+        expect(json['data']['attributes']['deck'].all? {
+          |card| card['attributes'].keys == [ "symbol", "number", "color", "shading" ]
```
Comment (efatsi) [inline spec/requests/games_spec.rb]: lil tip, you can use `=~` to match that the contents of an array are the same, even if their position is not. Not a problem for you here, but very handy when the order of an array potentially changes.

File: spec/models/game_spec.rb [hunk 12]
```diff
@@ -2,10 +2,16 @@
 
 RSpec.describe Game, type: :model do
 
+  subject { create :game }
+
   context "validations" do
```
Comment (ltk) [inline spec/models/game_spec.rb]: `validations` seems like a strange context for the below specs

File: spec/models/lobby_spec.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+RSpec.describe Lobby, type: :model do
+
+  subject { create :lobby }
+
+  context "validations" do
```
Comment (ltk) [inline spec/models/lobby_spec.rb]: same `validations` context comment here

File: lib/dealer.rb (hunk 7 above)
Comment (efatsi) [inline lib/dealer.rb]: AHA, you're shuffling an instance of cards to pass the Game for it's record of what the deck is, but the order never changes in here. so an array index lookup is equivalent to an id lookup. -> rather confusing. While it works, I'd love to see you make use of the `id` on Card with a detect lookup.

File: spec/requests/games_spec.rb [hunk 14]
```diff
@@ -2,24 +2,34 @@
...
       it "responds with game data" do
         expect(response).to have_http_status 200
         expect(json['data']['type']).to eq "games"
-        expect(json['data']['attributes']).to have_key("created_at")
+      end
+      it "responds with card data" do
+        expect(json['data']['attributes']).to have_key("board")
+        expect(json['data']['attributes']).to have_key("deck")
+        expect(json['data']['attributes']['board'].all? {
+          |card| card['attributes'].keys == [ "symbol", "number", "color", "shading" ]
```
Comment (ltk) [inline spec/requests/games_spec.rb]: seems like overkill to test all keys for all cards in the requests, without testing any values. Perhaps you could test one card's keys and values. or perhaps you could lean on your serializer unit specs for most of the actual response specifics.

File: spec/models/game_spec.rb [hunk 15]
```diff
@@ -2,10 +2,16 @@
...
 
-    it { is_expected.to have_many :users }
+    it { is_expected.to belong_to :lobby }
 
+    it "starts with 12 cards on the board" do
+      expect(subject.board.count).to eq 12
+      expect(subject.deck.count).to eq 69
```
Comment (efatsi) [inline spec/models/game_spec.rb]: shouldn't this be 81? 3^4

**Comment by ltk** [comment]
looking great! nice, simple stuff around card/deck management :+1:

File: app/controllers/api/v1/games_controller.rb (hunk 3 above)
Comment (connorlay) [inline app/controllers/api/v1/games_controller.rb]: What about a 422 Unprocessable Entity? [This blog post](https://localize-software.phraseapp.com/posts/best-practice-10-design-tips-for-apis/) suggests that status code for validation errors (which this is similar to).

File: app/controllers/api/v1/application_controller.rb (hunk 2 above)
Comment (connorlay) [inline app/controllers/api/v1/application_controller.rb]: I was trying to go for a simple `<noun>_error` naming convention, but I'm not sure if that makes more or less sense :no_mouth:

File: app/models/game.rb (hunk 6 above)
Comment (connorlay) [inline app/models/game.rb]: Yes! Spelling is hard :unamused:

File: app/models/game.rb (hunk 8 above)
Comment (connorlay) [inline app/models/game.rb]: I agree. That functionality belongs in `GameSerializer`, since it's role is to display a game (which this is all about anyway). I'll make the change in the other feature branch to avoid having to wrangle with fixing it here only to have major portions of `dealer` and `game` change.

File: lib/card.rb (hunk 9 above)
Comment (connorlay) [inline lib/card.rb]: I wanted to avoid explicitly defining the attributes in the card itself, which is probably emulating poodir a _little_ too much without considering the costs associated with `method_missing`. I already removed the `method_missing` override in the other feature branch since it didn't make much sense to me later on.

I toyed with the idea of making `Card` inherit from `OpenStruct` to make it explicit that card is just a data container. Thoughts?

File: app/models/game.rb (hunk 10 above)
Comment (connorlay) [inline app/models/game.rb]: In an integer field? I don't think I follow...

Regardless this method now looks like `12.times { game.board << game.deck.pop }` in the other feature branch since `#pop(n)` returns a new array, while `#pop` is destructive.

File: spec/requests/games_spec.rb (hunk 11 above)
Comment (connorlay) [inline spec/requests/games_spec.rb]: Ah! I remember `=~` from the regex chapter in wgr. I'll keep that in mind next time I'm comparing arrays.

File: spec/models/game_spec.rb (hunk 15 above)
Comment (connorlay) [inline spec/models/game_spec.rb]: There are 81 cards in total. 12 of them start on the board and the other 69 are in the deck.

File: app/controllers/api/v1/application_controller.rb (hunk 1 above)
Comment (connorlay) [inline app/controllers/api/v1/application_controller.rb]: I think I'll leave it there for now since that's where the rest of the HTTPS error responses are.

File: app/models/game.rb (hunk 10 above)
Comment (efatsi) [inline app/models/game.rb]: isn't `self.board=` setting the `board` database attribute (which looking at your schema is an integer field)?

File: spec/requests/games_spec.rb (hunk 14 above)
Comment (connorlay) [inline spec/requests/games_spec.rb]: I'm not exactly sure how to test values here. Since the board and deck are randomized, there is no guarantee which cards will be where. I _could_ change `game` before running the test and give it a specific `board` and `deck`, but I feel like that defeats the black box approach to integration testing.

I'll look into unit testing the serializer.

File: app/models/game.rb (hunk 10 above)
Comment (connorlay) [inline app/models/game.rb]:
~~~ruby
#/db/migrate/20150727171139_create_games.rb
t.integer :deck,  array: true, default: [], null: false
t.integer :board, array: true, default: [], null: false
~~~

`board` and `deck` are postgres integer arrays.

Comment (efatsi) [inline app/models/game.rb]: pardon my eyeballs. super cool.

File: app/controllers/api/v1/games_controller.rb (hunk 4 above)
Comment (efatsi) [inline app/controllers/api/v1/games_controller.rb]: bump

Comment (connorlay) [inline app/controllers/api/v1/games_controller.rb]: Just added it with a `check_user_membership` before filter.

File: lib/card.rb (hunk 9 above)
Comment (efatsi) [inline lib/card.rb]: I think this is a decent approach here. Only in rare cases will this stop the regular flow, you run into issues when this starts doing unexpected things, but this seems like a decently constrained use case.

Another more explicit option would be something like this:
```ruby
def initialize(id, attributes)
  @id = id
  @attributes = attributes
  define_attribute_methods
end

def define_attribute_methods
  @attributes.keys.each do |key|
    define_method key do
      @attributes[key]
    end
  end
end
```

Comment (ltk) [inline lib/card.rb]: if you choose to keep the `method_missing` override, you should also override `respond_to?`

Comment (connorlay) [inline lib/card.rb]:
~~~ruby
[9] pry(main)> card
=> #<Card:0x007feb6596e7d8 @attributes={:symbol=>:oval, :number=>:one, :color=>:purple, :shading=>:open}, @id=62>
[10] pry(main)> card.respond_to? :symbol
=> false
[11] pry(main)> card.symbol
=> :oval
~~~

Yeah I think so since this makes no sense. While `method_missing` is neat, I think the cost here is a little to high and this smells a lot like needless metaprogramming.

Comment (connorlay) [inline lib/card.rb]: I think it's fine for `Card` to have specific attributes. `Dealer` could have the `CARD_ATTRIBUTES` constant, so adding or removing attributes is an easy change localized to those two classes. I think what I'm really looking for is a nice wrapper around an attributes hash that also has an id.  I already wrote `Permutator` to be independent of the number of attributes so `Card` can really be represented any number of ways.

**Comment by efatsi** [comment]
:+1:
